File: algorithms/cpp/numberofAtoms/numberofAtoms.cpp
// Source: https://leetcode.com/problems/number-of-atoms/
// Author: Miao Zhang
// Date:   2021-03-04

class Solution {
public:
    string countOfAtoms(string formula) {
        int i = 0;
        string res;
        for (const auto& kv: countOfAtoms(formula, i)) {
            res += kv.first;
            if (kv.second > 1) res += std::to_string(kv.second);
        }
        return res;
    }

private:
    map<string, int> countOfAtoms(const string& formula, int& i) {
        map<string, int> counts;
        while (i != formula.length()) {
            if (formula[i] == '(') {
                const auto& tmp_counts = countOfAtoms(formula, ++i);
                int count = getCount(formula, i);
                for (const auto& kv: tmp_counts) {
                    counts[kv.first] += kv.second * count;
                }
            } else if (formula[i] == ')') {
                i++;
                return counts;
            } else {
                const string& name = getName(formula, i);
                counts[name] += getCount(formula, i);
            }
        }
        return counts;
    }

    string getName(const string& formula, int& i) {
        string name;
        while (isalpha(formula[i]) &&
               (name.empty() || islower(formula[i]))) {
            name += formula[i++];
        }
        return name;
    }

    int getCount(const string& formula, int& i) {
        string cnt_str;
        while (isdigit(formula[i])) cnt_str += formula[i++];
        return cnt_str.empty() ? 1 : std::stoi(cnt_str);
    }
};

File: algorithms/python/numberofAtoms/numberofAtoms.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/number-of-atoms/
# Author: Miao Zhang
# Date:   2021-03-04

class Solution:
    def countOfAtoms(self, formula: str) -> str:
        self.i = 0
        res = ''
        cnt = self.countAtoms(formula)
        for k in sorted(cnt.keys()):
            res += k
            if cnt[k] > 1:
                res += str(cnt[k])
        return res
    
    def countAtoms(self, formula: str) -> dict:
        counts = collections.defaultdict(int)
        while self.i != len(formula):
            if formula[self.i] == '(':
                self.i += 1
                tmp_counts = self.countAtoms(formula)
                count = self.getCount(formula)
                for k, v in tmp_counts.items():
                    counts[k] += v * count
            elif formula[self.i] == ')':
                self.i += 1
                return counts
            else:
                name = self.getName(formula)
                counts[name] += self.getCount(formula)
        return counts
    
    def getName(self, formula: str) -> str:
        name = ''
        while self.i < len(formula) and formula[self.i].isalpha() and (not name or formula[self.i].islower()):
            name += formula[self.i]
            self.i += 1
        return name
    
    def getCount(self, formula: str) -> int:
        cnt_str = ''
        while self.i < len(formula) and formula[self.i].isdigit():
            cnt_str += formula[self.i]
            self.i += 1
        return 1 if not cnt_str else int(cnt_str)
